File: experimental-plugins/agentic-workflows/src/agentic_workflows/logging/logger.py
"""NDJSON structured logging for workflows."""

from __future__ import annotations

import json
from dataclasses import asdict, dataclass
from datetime import datetime, timezone
from enum import Enum
from pathlib import Path
from typing import Any


class LogLevel(Enum):
    """Log severity levels."""

    CRITICAL = "Critical"
    ERROR = "Error"
    WARNING = "Warning"
    INFORMATION = "Information"


@dataclass
class LogEntry:
    """A single log entry."""

    timestamp: str
    level: str
    step: str
    message: str
    context: dict[str, Any] | None = None


class WorkflowLogger:
    """NDJSON logger for workflow execution."""

    def __init__(self, workflow_id: str, repo_root: Path | None = None):
        self.workflow_id = workflow_id
        if repo_root is None:
            repo_root = Path.cwd()
        self.log_path = repo_root / "agentic" / "workflows" / workflow_id / "logs.ndjson"
        self.log_path.parent.mkdir(parents=True, exist_ok=True)

    def log(
        self,
        level: LogLevel,
        step: str,
        message: str,
        context: dict[str, Any] | None = None,
    ) -> None:
        """Write a log entry."""
        entry = LogEntry(
            timestamp=datetime.now(timezone.utc).isoformat(),
            level=level.value,
            step=step,
            message=message,
            context=context,
        )

        with open(self.log_path, "a", encoding="utf-8") as f:
            f.write(json.dumps(asdict(entry)) + "\n")

    def critical(self, step: str, message: str, **context: Any) -> None:
        """Log a critical message."""
        self.log(LogLevel.CRITICAL, step, message, context or None)

    def error(self, step: str, message: str, **context: Any) -> None:
        """Log an error message."""
        self.log(LogLevel.ERROR, step, message, context or None)

    def warning(self, step: str, message: str, **context: Any) -> None:
        """Log a warning message."""
        self.log(LogLevel.WARNING, step, message, context or None)

    def info(self, step: str, message: str, **context: Any) -> None:
        """Log an informational message."""
        self.log(LogLevel.INFORMATION, step, message, context or None)


def read_logs(workflow_id: str, repo_root: Path | None = None) -> list[LogEntry]:
    """Read all log entries for a workflow."""
    if repo_root is None:
        repo_root = Path.cwd()
    log_path = repo_root / "agentic" / "workflows" / workflow_id / "logs.ndjson"

    if not log_path.exists():
        return []

    entries = []
    with open(log_path, encoding="utf-8") as f:
        for line in f:
            line = line.strip()
            if line:
                data = json.loads(line)
                entries.append(LogEntry(**data))
    return entries


def get_log_path(workflow_id: str, repo_root: Path | None = None) -> Path:
    """Get the path to the log file for a workflow."""
    if repo_root is None:
        repo_root = Path.cwd()
    return repo_root / "agentic" / "workflows" / workflow_id / "logs.ndjson"

File: src/logging/logger.ts
/** NDJSON structured logging for workflows. */

import { appendFileSync, existsSync, mkdirSync, readFileSync } from "node:fs";
import path from "node:path";

// --- Log levels ---

export const LOG_LEVEL = {
	CRITICAL: "Critical",
	ERROR: "Error",
	WARNING: "Warning",
	INFORMATION: "Information",
} as const;

export type LogLevel = (typeof LOG_LEVEL)[keyof typeof LOG_LEVEL];

// --- Log entry ---

export interface LogEntry {
	timestamp: string;
	level: string;
	step: string;
	message: string;
	context: Record<string, unknown> | null;
}

// --- Logger class ---

export class WorkflowLogger {
	readonly workflowId: string;
	readonly logPath: string;

	constructor(workflowId: string, outputDir: string) {
		this.workflowId = workflowId;
		this.logPath = path.join(outputDir, "logs.ndjson");
		mkdirSync(path.dirname(this.logPath), { recursive: true });
	}

	log(
		level: LogLevel,
		step: string,
		message: string,
		context?: Record<string, unknown> | null,
	): void {
		const entry: LogEntry = {
			timestamp: new Date().toISOString(),
			level,
			step,
			message,
			context: context ?? null,
		};
		appendFileSync(this.logPath, `${JSON.stringify(entry)}\n`, "utf-8");
	}

	critical(step: string, message: string, context?: Record<string, unknown>): void {
		this.log(LOG_LEVEL.CRITICAL, step, message, context || null);
	}

	error(step: string, message: string, context?: Record<string, unknown>): void {
		this.log(LOG_LEVEL.ERROR, step, message, context || null);
	}

	warning(step: string, message: string, context?: Record<string, unknown>): void {
		this.log(LOG_LEVEL.WARNING, step, message, context || null);
	}

	info(step: string, message: string, context?: Record<string, unknown>): void {
		this.log(LOG_LEVEL.INFORMATION, step, message, context || null);
	}
}

// --- Utility functions ---

export function readLogs(workflowId: string, outputDir: string): LogEntry[] {
	const logPath = getLogPath(workflowId, outputDir);
	if (!existsSync(logPath)) {
		return [];
	}

	const content = readFileSync(logPath, "utf-8");
	const entries: LogEntry[] = [];
	for (const line of content.split("\n")) {
		const trimmed = line.trim();
		if (trimmed) {
			try {
				entries.push(JSON.parse(trimmed) as LogEntry);
			} catch {
				// Skip corrupted log lines
			}
		}
	}
	return entries;
}

export function getLogPath(workflowId: string, outputDir: string): string {
	return path.join(outputDir, "logs.ndjson");
}
